lib/pbench/server/cache_manager.py

```diff
@@ -280,13 +281,12 @@ def extract(self, path: str) -> str:
             The named file as a string
         """
         try:
-            return (
-                tarfile.open(self.tarball_path, "r:*").extractfile(path).read().decode()
-            )
+            return tarfile.open(tarball_path, "r:*").extractfile(path).read().decode()
```

Comment:
So by using `fs-version-002` we won't break existing `fs-version-001` controller named based hierarchies.

Reply:
I'd been avoiding even thinking about departing that radically in the context of *this* PR.

But now that I think about it ... it might not be that bad, since the indexer now relies on the cache manager to look up a `Tarball` by resource ID and looks for the `unpacked` directory there. So does the inventory API; so it's possible that the disruption would be minimal. (All in all I'd still prefer to defer that to a separate PR, if nothing else because I'd like to get my #3194 on top of this.)

And regardless of this, I really don't want to defer capturing `dataset.metalog` and creation timestamp until we unpack. Once we graduate to a real cache manager this'll all be available dynamically as soon as we turn over control from `PUT` to cache manager. In the meantime, this seemed like a reasonable compromise. Of course if we'd rather use `tar` than `tarfile`, we could just do `tar -xfO {tarball} {tarball}/metadata.log` and capture the `stdout` stream.

Reply:
I'd like to remove all unpacking of the tar ball in the upload API.

The "created" field is not important, since the expiration is set based on the upload time.

If the "created" field is not there (yet) the UI does not have to display it.

That reduces the lifting being done during an already expensive API.

Then the unpack operation can add all that metadata to the dataset and it will show up in the UI.

Once we have a message passing architecture, then we can hide all these steps such that the data set can be in a "processing" mode in the background if the UI catches it in that state.

So I'd like to see a PR remove the unpacking for metadata behavior in the upload API, another PR remove the cache manager use of controller and move to an MD5 hierarchy, and then have this change land.

Landing this now expands the use of the metadata in the upload API, which I don't want to have to back out later.

Reply:
1. I'd agree that the `created` column in `Dataset` is no longer critical since I started capturing the entire `metadata.log`; I didn't have that originally and I'd wanted to retain that information. I'd basically tried to avoid random DB schema changes, partly because they tend to ripple through the code and tests. I wouldn't object to removing the column. If I did that, I'd probably move `metalog` from a `Metadata` key inline into `Dataset` as a `JSON` column. I might do `tarball-path` as well, since that'll be our S3 URI handle eventually. (I might rename it, too.)
2. However, having metadata not present for some arbitrary time after upload is inconsistent and awkward. I really don't like that idea, and I think it significantly reduces the value of metadata in general.
3. Removing the controller hierarchy in the filesystem makes a lot of sense. I'd been hoping we'd be able to remove filesystem dependencies with cache manager before we played with that; but we may actually already be there, between `Sync` and `CacheManager` APIs. But the way Riya coded it it's just another access to the information that we've already been capturing for a year now; and I really, really, dislike the idea of dropping the capture of `metadata.log`. We can argue separately about destroying metadata, but I don't see any sense in trying to make this PR suddenly dependent on unrelated changes in that area.

Reply:
Pete's always talked about the performance costs of `tarfile`; and it's not that I ever doubted that, but I've always been curious about quantifying the difference.

I wrote a little script to measure and compare average costs of directly extracting `metadata.log` via `tarfile.open().extract()` vs `subprocess.run('tar xf0')`. This  run points at the functional test tarballs (which are small, to be sure), measuring and averaging 500 iterations of each mechanism for each tarball.

As you can see, the actual times are tenths or hundredths of a second, and `tar` generally shows around 50% advantage here. Interesting.

__Anyway, this generates some data, but leaves the essential question of how critical this overhead is right now and how to weigh the tradeoffs. I kinda think that having the metadata available from upload is probably worth the 1/10 second it takes. And having extracted the metadata for `dataset.metalog`, using it again for the controller is "free". If the alternative is taking time to re-implement the file system hierarchy maybe getting the controller is an acceptable tradeoff now?__

```
Starting fio_rw_2018.02.01T22.40.57 (2.2 MB)
	 average tarfile extract time: 0.0736548
	 average tar extract time: 0.0331896
Starting linpack_mock_2020.02.28T19.10.55 (20.8 kB)
	 average tarfile extract time: 0.010313
	 average tar extract time: 0.00819257
Starting pbench-user-benchmark_example-vmstat_2018.10.24T14.38.18 (31.4 kB)
	 average tarfile extract time: 0.0167387
	 average tar extract time: 0.0106804
Starting trafficgen_basic-forwarding-example_tg:trex-profile_pf:forwarding_test.json_ml:5_tt:bs__2019-08-27T14:58:38 (1.0 MB)
	 average tarfile extract time: 0.268283
	 average tar extract time: 0.135689
Starting uperf_rhel8.1_4.18.0-107.el8_snap4_25gb_virt_2019.06.21T01.28.57 (8.4 MB)
	 average tarfile extract time: 0.159724
	 average tar extract time: 0.0774569
Tar wins 0.053 to 0.106 by 0.0527: 49.84%
```

Interesting trivia: on some smaller tarfiles (e.g. the ones on intlab-002 under the `ff00cfa30cd2` controller), the `tarfile` package actually wins by about 18% (a margin of 0.00026 seconds) but the largest file easily available (gandalf's fio run at 21.4 Mb) shows a 56% advantage to `tar`. Here, using `tarfile` adds about 0.17 seconds to `PUT` (switching to `tar` would cut that to 0.07).

I could easily run this on our 0.69 server, and it'd be interesting to see how this escalates on "big" datasets; but I seem to have lost my keys.

Reply:
I don't agree with this path of using `tarfile` or of unpacking the tarball on the Upload API.  But since others approve and do not seem concerned about it, I'll resolve my comment and withdraw my review.